tests/difference/core/quint_model/ccv_test.qnt

```diff
@@ -0,0 +1,255 @@
+
+    // negative voting powers give an error
+    run VotingPowerNegativeTest =
+    {
+        votingPowerChange(
+            GetEmptyProtocolState,
+            "validator",
+            -1
+        ).hasError
+    }
+
+    run VotingPowerOkTest =
+    {
+        val result = votingPowerChange(
+            GetEmptyProtocolState,
+            "validator",
+            5
+        )
+        not(result.hasError) and
+        result.newState.providerState.chainState.currentValidatorSet.keys().contains("validator") and
+        result.newState.providerState.chainState.currentValidatorSet.get("validator") == 5
+    }
+
+    // voting power of 0 is allowed and applied as usual
+    run VotingPowerZeroTest =
+    {
+        val tmpResult = votingPowerChange(
+            GetEmptyProtocolState,
+            "validator",
+            5
+        )
+        val finalResult = votingPowerChange(
+            tmpResult.newState,
+            "validator",
+            0
+        )
+        not(finalResult.hasError) and
+        finalResult.newState.providerState.chainState.currentValidatorSet.get("validator") == 0
+    }
+
+    run VotingPowerSetsFlagTest =
+    {
+        // change voting power
+        val tmpResult = votingPowerChange(
+            GetEmptyProtocolState,
+            "validator",
+            5
+        )
+        // but then set it back to the initial value
+        val finalResult = votingPowerChange(
+            tmpResult.newState,
+            "validator",
+            0
+        )
+        // still should set the flag
+        not(finalResult.hasError) and
+        finalResult.newState.providerState.providerValidatorSetChangedInThisBlock
+    }
+
+
+    // make sure that VotingPowerChange ONLY changes the current validator set, not the history
+    run VotingPowerChangeDoesNotChangeHistoryTest =
+    {
+        val result = votingPowerChange(
+            GetEmptyProtocolState,
+            "validator",
+            0
+        )
+        not(result.hasError) and
+        result.newState.providerState.chainState.votingPowerHistory == List()
+    }
+
+    // Defines a test state to test the deliverPacketToProvider function against.
+    pure val _DeliverPacketToProvider_TestState =
+        val currentState = GetEmptyProtocolState
+        val sender = "sender"
+        val providerState = currentState.providerState
+        val consumerState = GetEmptyConsumerState
+        // add the consumer to the consumerStates
+        val consumerStates = currentState.consumerStates.put(sender, consumerState)
+        val providerState2 = providerState.with(
+            "consumerStatus", providerState.consumerStatus.put(sender, RUNNING)
+        )
+        val providerState3 = providerState2.with(
+            "outstandingPacketsToConsumer", providerState2.outstandingPacketsToConsumer.put(sender, List({
```

Comment:
This was initially a bit confusing because the test state is named `_DeliverPacketToProvider_` but the initial state is having the provider ready to send a packet to the consumer.

Reply:
I think it's okay if it's confusing for the first few lines :D